cli/src/commands/fix.rs

```diff
@@ -328,3 +390,126 @@ fn run_tool(
+    matcher: Box<dyn Matcher>,
+    // TODO: Store the `name` field here and print it with the command's stderr, to clearly
+    // associate any errors/warnings with the tool and its configuration entry.
+}
+
+/// Represents the `fix.tools` config table.
+struct ToolsConfig {
+    /// Some tools, stored in the order they will be executed if more than one
+    /// of them matches the same file.
+    tools: Vec<ToolConfig>,
+}
+
+/// Parses the `fix.tools` config table.
+///
+/// Parses the deprecated `fix.tool-command` config as if it was the first entry
+/// in `fix.tools`.
+///
+/// Fails if any of the commands or patterns are obviously unusable, but does
+/// not check for issues that might still occur later like missing executables.
+/// This is a place where we could fail earlier in some cases, though.
+fn get_tools_config(
+    workspace_command: &WorkspaceCommandHelper,
+    config: &config::Config,
+) -> Result<ToolsConfig, CommandError> {
+    let mut tools_config = ToolsConfig { tools: Vec::new() };
+    if let Ok(tool_command) = config.get("fix.tool-command") {
+        // This doesn't change the displayed indices of the `fix.tools` definitions, and
+        // doesn't have a `name` that could conflict with them. That would matter more
+        // if we already had better error handling that made use of the `name`.
+        tools_config.tools.push(ToolConfig {
+            command: tool_command,
+            matcher: Box::new(EverythingMatcher),
+        });
+    }
+    if let Ok(tool_config_array) = config.get_array("fix.tools") {
+        let mut tool_name_first_seen_index = HashMap::new();
+        let mut tools: Vec<ToolConfig> = tool_config_array
+            .into_iter()
+            .enumerate()
+            .map(|(index, val)| -> Result<ToolConfig, CommandError> {
+                let table = val.into_table()?;
+                let name = table
+                    .get("name")
+                    .ok_or(config::ConfigError::Message(format!(
+                        "`fix.tools` entry at index {} is missing required field `name`.",
+                        index
+                    )))?
+                    .to_string();
+                if let Some(first_index) = tool_name_first_seen_index.insert(name.clone(), index) {
+                    return Err(config::ConfigError::Message(format!(
+                        "`fix.tools` entries at indices {} and {} have the same `name`: {}",
+                        first_index, index, name
+                    ))
+                    .into());
+                }
+                let patterns: Vec<String> = table
+                    .get("patterns")
+                    .ok_or(config::ConfigError::Message(format!(
+                        "`fix.tools` entry `{}` at index {} is missing required field `patterns`.",
+                        name, index
+                    )))?
+                    .clone()
+                    .into_array()?
```

Comment:
nit: `.try_deserialize()` can be used to map `config::Value` to arbitrary type.

We can also define intermediate config struct (like `ExternalMergeTool`) and deserialize into it.